src/core/manager-json.c

```diff
@@ -0,0 +1,202 @@
+                        SD_JSON_BUILD_PAIR_CALLBACK("defaultLimitLOCKS", rlimit_build_json, m->defaults.rlimit[RLIMIT_LOCKS]),
+                        SD_JSON_BUILD_PAIR_CALLBACK("defaultLimitSIGPENDING", rlimit_build_json, m->defaults.rlimit[RLIMIT_SIGPENDING]),
+                        SD_JSON_BUILD_PAIR_CALLBACK("defaultLimitMSGQUEUE", rlimit_build_json, m->defaults.rlimit[RLIMIT_MSGQUEUE]),
+                        SD_JSON_BUILD_PAIR_CALLBACK("defaultLimitNICE", rlimit_build_json, m->defaults.rlimit[RLIMIT_NICE]),
+                        SD_JSON_BUILD_PAIR_CALLBACK("defaultLimitRTPRIO", rlimit_build_json, m->defaults.rlimit[RLIMIT_RTPRIO]),
+                        SD_JSON_BUILD_PAIR_CALLBACK("defaultLimitRTTIME", rlimit_build_json, m->defaults.rlimit[RLIMIT_RTTIME]),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("defaultTasksMax", cgroup_tasks_max_resolve(&m->defaults.tasks_max)),
+                        JSON_BUILD_PAIR_FINITE_USEC("defaultMemoryPressureThresholdUSec", m->defaults.memory_pressure_threshold_usec),
+                        SD_JSON_BUILD_PAIR_STRING("defaultMemoryPressureWatch", cgroup_pressure_watch_to_string(m->defaults.memory_pressure_watch)),
+                        JSON_BUILD_PAIR_FINITE_USEC("timerSlackNSec", (uint64_t) prctl(PR_GET_TIMERSLACK)),
+                        SD_JSON_BUILD_PAIR_STRING("defaultOOMPolicy", oom_policy_to_string(m->defaults.oom_policy)),
+                        SD_JSON_BUILD_PAIR_INTEGER("defaultOOMScoreAdjust", m->defaults.oom_score_adjust),
+                        SD_JSON_BUILD_PAIR_STRING("ctrlAltDelBurstAction", emergency_action_to_string(m->cad_burst_action)));
+}
+
+static int log_level_build_json(sd_json_variant **ret, const char *name, void *userdata) {
+        _cleanup_free_ char *t = NULL;
+        int r;
+
+        assert(ret);
+
+        r = log_level_to_string_alloc(log_get_max_level(), &t);
+        if (r < 0)
+                return r;
+
+        return sd_json_variant_new_string(ret, t);
+}
+
+static int manager_environment_build_json(sd_json_variant **ret, const char *name, void *userdata) {
+        _cleanup_strv_free_ char **l = NULL;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        Manager *m = ASSERT_PTR(userdata);
+        int r;
+
+        assert(ret);
+
+        r = manager_get_effective_environment(m, &l);
+        if (r < 0)
+                return r;
+
+        STRV_FOREACH(s, l) {
+                _cleanup_free_ char *key = NULL;
+                const char *p = *s;
+
+                r = extract_first_word(&p, &key, "=", 0);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        return -EINVAL;
+
+                r = sd_json_variant_set_field_string(&v, key, *s);
+                if (r < 0)
+                        return r;
+        }
```

Comment:
for user records i opted to encode env var blocks as string array, without trying to split them up at "=". i think we should stick to that here too.

Reply:
also, the last argument to sd_json_variant_set_field_string() probably should have been `p`, no?